Enabling OIDC causes all robot accounts to generate error: failed to verify secret, username: robot$gitlab, error: failed to get oidc user info, error: <QuerySeter> no row

It appears when OIDC is enabled the auth goes through OIDC even if the account is a robot account with no OIDC. This causes the logs to be inundated with this error message:

`2021-07-02T14:55:11Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="a2af154dd57eca5e5a05a8ac4012a311"]: failed to verify secret, username: robot$gitlab, error: failed to get oidc user info, error: <QuerySeter> no row `

Since robot accounts cannot even be linked to OIDC accounts is there some sort of way to stop these log messages and errors?

I am using v2.3.0. OIDC settings are:

```
OIDC Provider Name : azuread
OIDC Endpoint: https://login.microsoftonline.com/CENSORED/v2.0
OIDC Client ID: CENSORED
OIDC Client Secret: CENSORED
Group Claim Name: groups
OIDC Admin Group: CENSORED
OIDC Scope: openid,email,profile,offline_access
Verify Certificate: On
Automatic Onboarding: On
Username Claim: email
```

Some more info from debug logs:

```
2021-07-03T11:37:41Z [DEBUG] [/pkg/oidc/secret.go:75]: Verifying the secret for user: robot$gitlab
2021-07-03T11:37:41Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="1135d5b3c15180380de126e5d2617cc9"]: failed to verify secret, username: robot$gitlab, error: failed to get oidc user info, error: <QuerySeter> no row found
2021-07-03T11:37:41Z [ERROR] [/pkg/token/token.go:66]: parse token error, token contains an invalid number of segments
2021-07-03T11:37:41Z [DEBUG] [/server/middleware/security/robot.go:49][requestID="1135d5b3c15180380de126e5d2617cc9"]: failed to decrypt robot token of v1 robot: robot$gitlab, as: token contains an invalid number of segments
```


Seems it tries to verify the robot token via oidc despite it not being oidc and then  it fails to do so and then the token it generates is invalid.

so at first I thought the logs were relevant, but the working setup logs exactly the same lines, even though auth works:
```
core 2021-07-05T20:15:40Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="de6b09e5-ce00-4369-a4d6-4a08054a72ce"]: failed to verify secret, username: robot$xxxxxx, error: failed to get oidc user info, error: <QuerySeter> no row found                                                                                            │
core 2021-07-05T20:15:40Z [ERROR] [/pkg/token/token.go:66]: parse token error, token contains an invalid number of segments
```

The cluster that is failing auth shows nothing different:
```
core 2021-07-05T20:19:07Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="4a41369c-21e6-4779-927a-45025a9d90c8"]: failed to verify secret, username: robot$team-otomi+kubernetes, error: failed to get oidc user info, error: <QuerySeter> no row found                                                                                                  │
core 2021-07-05T20:19:07Z [ERROR] [/pkg/token/token.go:66]: parse token error, token contains an invalid number of segments
```
The client that fails:
```
Error response from daemon: login attempt to https://harbor.xxxxxxx/v2/ failed with status: 401 Unauthorized
```

Seems I still see these after the adjustment made above:

`021-07-06T12:22:07Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="9d5bf08d0598b184a4a9f7ce39f65f0c"]: failed to verify secret, username: robot$gitlab, error: failed to get oidc user info, error: <QuerySeter> no row found`

also still seeing these:

`2021-07-06T12:22:07Z [DEBUG] [/lib/http/error.go:59]: {"errors":[{"code":"UNAUTHORIZED","message":"authorize header needed to send HEAD to repository: authorize header needed to send HEAD to repository"}]}`

Hello, did anyone find any solution for this issue?

I just upgraded from 2.2.1 to 2.3.1 and see same error. I am not able to login to docker using robot account. 

Jul 26 16:22:22  core[2519]: 2021-07-26T16:22:22Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="092sfa74-c11b-4966-92a9-e6alleda2eae"]: failed to verify secret, username: robot$testupgrade, error: failed to get oidc user info, error: <QuerySeter> no row found
Jul 26 16:22:22 core[2519]: 2021-07-26T16:22:22Z [ERROR] [/pkg/token/token.go:66]: parse token error, token contains an invalid number of segments
Jul 26 16:22:22 core[2519]: 2021-07-26T16:22:22Z [ERROR] [/server/middleware/security/robot2.go:47][requestID="092sfa74-c11b-4966-92a9-e6alleda2eae"]: failed to authenticate robot account: robot$testupgrade
Jul 26 16:22:22 core[2519]: 2021-07-26T16:22:22Z [ERROR] [/core/auth/authenticator.go:264]: Failed to get user from DB, username: robot$testupgrade, error: user robot$testupgrade not found

So more digging this error could be solved by adding a check to this method to return nil if the username prefix is the same as the robot prefix: 

`2021-08-24T11:15:42Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="f7b12e7636f6e9fd320dd6e4ac518d77"]: failed to verify secret, username: ROBOTACCOUNT, error: failed to get oidc user info, error: <QuerySeter> no row found`

This would eliminate one of the big error logs we see.

Still seeing this with 2.3.3, at least when doing a `helm chart push` using a robot account.

```
harbor-test-core-fd897d4f8-z9p8s core 2021-09-28T21:36:49Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="eb2db930c17222d5b43cb97637254b1d"]: failed to verify secret, username: robot+helm+atc135, error: failed to get oidc user info, error: <QuerySeter> no row found
harbor-test-core-fd897d4f8-z9p8s core 2021-09-28T21:36:49Z [INFO] [/server/middleware/security/robot2.go:74][requestID="eb2db930c17222d5b43cb97637254b1d"]: a robot2 security context generated for request GET /service/token
```

The client then shows:

```
Error: unexpected response: 401 Unauthorized
```

Did anyone resolve this? I am seeing same issue in 2.3.4 version. It still thinks robot user as oidc user and tries to get info from DB.

Nov 12 18:16:19  core[2641]: 2021-11-12T18:16:19Z [ERROR] [/server/middleware/security/robot2.go:47][requestID="9ded48d7-5d7b-44350-aeb9-4b074796de2"]: failed to authenticate robot account: robot$testupgrade

Nov 12 18:16:19  core[2641]: 2021-11-12T18:16:19Z [ERROR] [/core/auth/authenticator.go:262]: Failed to get user from DB, username: robot$testupgrade, error: user robot$testupgrade not found

Nov 12 18:16:19 [2641]: 2021-11-12T18:16:19Z [ERROR] [/server/middleware/security/basic_auth.go:40][requestID="9ded48d7-5d7b-44350-aeb9-4b074796de2"]: failed to authenticate robot$testupgrade: not supported

Did anyone resolve this? I am seeing below for 2.3.4 harbor version.

Nov 12 18:25:54  core[2641]: 2021-11-12T18:25:54Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="07a74387-4cc5-4d94-bde9-b3dge0509939"]: failed to verify secret, username: robot$testupgrade, error: failed to get oidc user info, error: <QuerySeter> no row found

Nov 12 18:25:54  core[2641]: 2021-11-12T18:25:54Z [ERROR] [/pkg/token/token.go:66]: parse token error, token contains an invalid number of segments

Nov 12 18:25:54  core[2641]: 2021-11-12T18:25:54Z [INFO] [/server/middleware/security/robot2.go:74][requestID="07a74387-4cc5-4d94-bde9-b3dge0509939"]: a robot2 security context generated for request GET /service/token

Not working for us either with Version v2.4.0-d4affc2e

`2021-12-27T16:07:58Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="db599dc81b53bfb8b80d0e1816bd4757"]: failed to verify secret, username: robot$project+helm-push, error: failed to get oidc user info, error: <QuerySeter> no row found`

Same issue:
Version v2.4.1-c4b06d79

```
Dec 31 07:58:03 172.19.0.1 core[2343]: 2021-12-31T07:58:03Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="0a5dd53c-531c-47e9-a446-ea3af701d903"]: failed to verify secret, username: robot$up-node, error: failed to get oidc user info, error: <QuerySeter> no row found
Dec 31 07:58:03 172.19.0.1 core[2343]: 2021-12-31T07:58:03Z [ERROR] [/server/middleware/security/basic_auth.go:40][requestID="0a5dd53c-531c-47e9-a446-ea3af701d903"]: failed to authenticate robot$up-node: not supported
```

We upgraded from 1.8 to latest version and we are in the same situation. This happens in automated robot accounts like trivy one too. 

````harbor-core-68f446c9bf-pjm8v core 2022-02-04T06:55:09Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="1a3d6e25-9104-44f7-99a7-1dc56a507c62"]: failed to verify secret, username: robot$xxx+Trivy-cbdab39d-8585-11ec-aabe-765eaaf3b54e, error: failed to get oidc user info, error: <QuerySeter> no row found```

We've been receiving these same error messages for some time now, however our GitHub Actions image publishing was working up until last Thursday when our Harbor Helm release auto-updated. Now, all we are seeing is:
`[ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="xxxxxxxx"]: failed to verify secret, username: xxxx, error: failed to get oidc user info, error: <QuerySeter> no row`

I'm getting these errors on a fresh install of 2.5.3. OIDC is configured, and seems to be work ok. Created a robot account to do replication from another harbor install and seeing a lot of these errors. 

```
NAME  	NAMESPACE    	REVISION	UPDATED                             	STATUS  	CHART       	APP VERSION
marina	harbor-marina	1       	2022-07-29 19:28:33.161549 -0600 MDT	deployed	harbor-1.9.3	2.5.3
```


Also, I'm unable to login using docker, but I can login to the UI. I copied the token from my profile.
```
docker login marina.[redacted] -u ned_hanks -p v5HIeUuUx
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Error response from daemon: Get "https://marina.octanner.io/v2/": unauthorized: authentication required
```
```
2022-08-12T23:27:03Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="03bc7f14-0fa2-4eda-9c16-eb1d9877043a"]: failed to verify secret, username: ned_hanks, error: failed to get oidc user info, error: <QuerySeter> no row found
2022-08-12T23:27:03Z [ERROR] [/server/middleware/security/basic_auth.go:72][client IP="10.2.48.0, 10.2.9.212" requestID="03bc7f14-0fa2-4eda-9c16-eb1d9877043a" user agent="docker/20.10.12 go/go1.16.12 git-commit/459d0df kernel/5.10.76-linuxkit os/linux arch/amd64 UpstreamClient(Docker-Client/20.10.12 \(darwin\))"]: failed to authenticate user:ned_hanks, error:not supported
```

I'm still running into this issue. Deploying Harbor using the Helm chart version v1.10.2

```
[ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="552f33a7-4a03-4b04-8782-f5aa6152b578"]: failed to verify secret, username: robot$account, error: failed to get oidc user info, error: <QuerySeter> no row found
[INFO] [/server/middleware/security/robot.go:71][requestID="552f33a7-4a03-4b04-8782-f5aa6152b578"]: a robot security context generated for request GET /service/token
```

I also see this on v1.10.2, but with slightly different errors:

```
2022-11-21T07:56:55Z [ERROR] [/server/middleware/security/basic_auth.go:72][client IP="10.244.0.70:55126" requestID="4251fe16-8e20-4c74-ad08-774836c87951" user agent="curl/7.72.0-DEV"]: failed to authenticate user:admin, error:Failed to authenticate user, due to error 'Invalid credentials'
2022-11-21T07:56:57Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="70a43bf6-3933-4c4a-b79d-9b93ab2a4485"]: failed to verify secret, username: admin, error: failed to get oidc user info, error: <QuerySeter> no row found
2022-11-21T07:56:57Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="3eb4527f-1039-463f-ba57-4d2f10a5122c"]: failed to verify secret, username: admin, error: failed to get oidc user info, error: <QuerySeter> no row found
2022-11-21T07:56:57Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="72f15e3f-8f9d-45c6-8e5e-e257e4e8f216"]: failed to verify secret, username: admin, error: failed to get oidc user info, error: <QuerySeter> no row found
```

Mabye that also happens due to migration from v.1.10.1 -> v1.10.2.

Same here, waiting for this fix to be out once 
```shell
2023-01-30T13:21:56.445777863Z 2023-01-30T13:21:56Z [INFO] [/jobservice/logger/service.go:63]: Found export data cleanup job with schedule id : 1
2023-01-30T13:22:56.435408139Z 2023-01-30T13:22:56Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'AuditLog' on topic 'CREATE_PROJECT': ID-2 Name-gitlab Operator-hrouineb OccurAt-2023-01-30 13:22:56
2023-01-30T13:24:57.636334678Z 2023-01-30T13:24:57Z [INFO] [/controller/registry/controller.go:222]: Start regular health check for registries with interval 5m0s
2023-01-30T13:26:03.476178027Z 2023-01-30T13:26:03Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="2115c3f2-a75e-4742-aeed-83d52c917773"]: failed to verify secret, username: robot$gitlab+yinn, error: failed to get oidc user info, error: <QuerySeter> no row found
2023-01-30T13:26:03.483522200Z 2023-01-30T13:26:03Z [INFO] [/server/middleware/security/robot.go:71][requestID="2115c3f2-a75e-4742-aeed-83d52c917773"]: a robot security context generated for request GET /service/token
2023-01-30T13:28:44.937218308Z 2023-01-30T13:28:44Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="15cd5e03-4df9-4d74-83b9-9a01eec5de8d"]: failed to verify secret, username: robot$gitlab+yinn, error: failed to get oidc user info, error: <QuerySeter> no row found
2023-01-30T13:28:44.947952864Z 2023-01-30T13:28:44Z [INFO] [/server/middleware/security/robot.go:71][requestID="15cd5e03-4df9-4d74-83b9-9a01eec5de8d"]: a robot security context generated for request GET /service/token
2023-01-30T13:33:43.333409691Z 2023-01-30T13:33:43Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="206acf81-b59b-4564-b647-91f55b3bba86"]: failed to verify secret, username: robot$gitlab+yinn, error: failed to get oidc user info, error: <QuerySeter> no row found
2023-01-30T13:33:43.342219923Z 2023-01-30T13:33:43Z [INFO] [/server/middleware/security/robot.go:71][requestID="206acf81-b59b-4564-b647-91f55b3bba86"]: a robot security context generated for request GET /service/token
2023-01-30T13:34:02.902550312Z 2023-01-30T13:34:02Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="487cec82-70cf-438f-9638-11b3fe2d5eb4"]: failed to verify secret, username: robot$gitlab+yinn, error: failed to get oidc user info, error: <QuerySeter> no row found
2023-01-30T13:34:02.909669962Z 2023-01-30T13:34:02Z [INFO] [/server/middleware/security/robot.go:71][requestID="487cec82-70cf-438f-9638-11b3fe2d5eb4"]: a robot security context generated for request GET /service/token
2023-01-30T13:38:45.055944272Z 2023-01-30T13:38:45Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="95eb830b-e4df-4929-9700-dd5d9e5ea436"]: failed to verify secret, username: robot$hisqool, error: failed to get oidc user info, error: <QuerySeter> no row found
2023-01-30T13:38:45.063001194Z 2023-01-30T13:38:45Z [INFO] [/server/middleware/security/robot.go:71][requestID="95eb830b-e4df-4929-9700-dd5d9e5ea436"]: a robot security context generated for request GET /service/token

```

--- a/src/server/middleware/security/oidc_cli.go
+++ b/src/server/middleware/security/oidc_cli.go
@@ -26,6 +26,7 @@ import (
 	"github.com/goharbor/harbor/src/common/security/local"
 	"github.com/goharbor/harbor/src/controller/user"
 	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/config"
 	"github.com/goharbor/harbor/src/lib/log"
 	"github.com/goharbor/harbor/src/pkg/oidc"
 )
@@ -57,6 +58,11 @@ func (o *oidcCli) Generate(req *http.Request) security.Context {
 	if !o.valid(req) {
 		return nil
 	}
+
+	if strings.HasPrefix(username, config.RobotPrefix(ctx)) {
+		return nil
+	}
+
 	info, err := oidc.VerifySecret(ctx, username, secret)
 	if err != nil {
 		logger.Errorf("failed to verify secret, username: %s, error: %v", username, err)

